### PR TITLE
Fix LSTM and GRU cuDNN kernel failure for RaggedTensors.

### PR DESCRIPTION
**Pull request by foxik**
The `keras.layers.LSTM` and `keras.layers.GRU` layers currently do not use cuDNN kernel for `tf.RaggedTensor`s, as reported by https://github.com/tensorflow/tensorflow/issues/48838 and also as documented in the code at
- https://github.com/keras-team/keras/blob/5d2993b78603bda911f0a4912d08b1e5f36eb633/keras/layers/recurrent_v2.py#L424-L426
- https://github.com/keras-team/keras/blob/5d2993b78603bda911f0a4912d08b1e5f36eb633/keras/layers/recurrent_v2.py#L1153-L1155

I have traced the problem to https://github.com/keras-team/keras/blob/5d2993b78603bda911f0a4912d08b1e5f36eb633/keras/layers/recurrent_v2.py#L616-L623 -- the problem is that even if `mask` can be `None`, `sequence_lengths` might not be `None` -- and in that case the CudnnRNNV3 is used https://github.com/keras-team/keras/blob/5d2993b78603bda911f0a4912d08b1e5f36eb633/keras/layers/recurrent_v2.py#L655-L664 and the inputs **should not** be transposed.

This condition (`mask=None`, `sequence_lengths != None`) is actually forgotten in several places in `gpu_gru` and `gpu_lstm` -- also on https://github.com/keras-team/keras/blob/5d2993b78603bda911f0a4912d08b1e5f36eb633/keras/layers/recurrent_v2.py#L686-L687 and then on https://github.com/keras-team/keras/blob/5d2993b78603bda911f0a4912d08b1e5f36eb633/keras/layers/recurrent_v2.py#L690-L697. I therefore provide a patch, which:
- moves the copy of `mask` to `sequence_lengths` https://github.com/keras-team/keras/blob/5d2993b78603bda911f0a4912d08b1e5f36eb633/keras/layers/recurrent_v2.py#L652-L653 to the beginning of the function
- then checks only for `sequence_lengths` being `None` or not -- this way both cases (either `mask` or `sequence_lengths` being non-None) are handled consistently.

I also removed the test disallowing using the cuDNN kernel for `tf.keras.layers.{LSTM/GRU}` and removed the reference to the internal bug: https://github.com/keras-team/keras/blob/5d2993b78603bda911f0a4912d08b1e5f36eb633/keras/layers/recurrent_v2.py#L1153-L1154 https://github.com/keras-team/keras/blob/5d2993b78603bda911f0a4912d08b1e5f36eb633/keras/layers/recurrent_v2.py#L424-L425 . But maybe this needs to be handled also internally in Google somehow?

### CODE REVIEW

**Comment by foxik** [comment]
Gentle bump after 10 days of inactivity.

Could I maybe ask @qlzh727 for a review or feedback? Thanks!

**Review by fchollet — COMMENTED** [review]
Thanks for the PR! Please add a unit test for this change.

**Comment by foxik** [comment]
Thanks for the feedback!

The PR only changes a specific implementation used for LSTM and GRU on a GPU for RaggedTensors -- as far as I know, it is not straightforward to write a test that the cuDNN implementation is now used. Furthermore, there is a test checking that the computation with RaggedTensors finished (because it was failing before), see https://github.com/keras-team/keras/blob/5d2993b78603bda911f0a4912d08b1e5f36eb633/keras/layers/recurrent_v2_test.py#L112-L120

What I could imagine is to add a test which would verify that LSTM and GRU gives the same results, when considering sequences of different lengths via:
- masking
- RaggedTensors
- masking on a GPU (if GPU available)
- RaggedTensors on a GPU (if GPU available)

This way, we would get a reasonable test checking that all both masking/RaggedTensors and CPU/GPU implementations give the same results.

Do you think it is reasonable? Or do you have other test in mind?

Cheers!

**Comment by fchollet** [comment]
> Do you think it is reasonable? Or do you have other test in mind?

Yes, such a test would be perfect. Thank you.

**Comment by foxik** [comment]
Thanks for the response, I just added a test which compares the CPU/GPU and masked/ragged tensors, on both LSTM and GRU. Do you think it is fine like this?

BTW, the applied patch makes the ragged tensors _considerably_ faster on a GPU -- for a POS tagger with a sentence-level LSTM with 256 units (completed version of this template https://github.com/ufal/npfl114/blob/master/labs/08/tagger_we.py), the:
- patched GPU implementation takes 14ms/training step
- unpatched GPU implementation takes 140ms/training step, so 10 times slower
- for reference, CPU implementation takes 335ms/training step

Cheers!

**Review by fchollet — APPROVED** [review]
LGTM, thank you for the contribution 👍

**Comment by fchollet** [comment]
> the applied patch makes the ragged tensors considerably faster on a GPU 

That's awesome!

**Comment by fchollet** [comment]
We weren't able to merge yet due to a failing test:

```
RNNV2Test.test_compare_ragged_with_masks1_v1_session (<class 'keras.layers.recurrent_v2.GRU'>)
```

Can you please take a look?

**Comment by foxik** [comment]
I will gladly help -- but I cannot see any mistake. In the visible build status, I see 1 failing and 5 successful checks.

- the successful checks include Ubuntu-CPU and Ubuntu-GPU, both working fine (including the `recurrent_v2_test{,_gpu}`)
  - the Ubuntu-CPU logs are at https://source.cloud.google.com/results/invocations/33f62559-64e7-45f4-bae8-2cb3ea3a3694/targets/keras%2Fgithub%2Fubuntu%2Fgpu%2Fpresubmit/log
  - the Ubuntu-CPU logs are at https://source.cloud.google.com/results/invocations/0de42046-6d92-4171-bdf9-d022a86be1a1/targets/keras%2Fgithub%2Fubuntu%2Fcpu%2Fpresubmit/log

- the unsuccessful run `feedback/copybara` points to Google-internal cl/418079492, which I cannot see

- the tests worked fine on my machine (tf-nightly, CUDA 11.2, cuDNN 8.1), using GTX 3090 and I just rechecked with 1080 Ti. However, I see that it tested only
  - RNNV2Test.test_compare_ragged_with_masks0_v2_eager
  - RNNV2Test.test_compare_ragged_with_masks1_v2_eager
  - RNNV2Test.test_compare_ragged_with_masks0_v2_function
  - RNNV2Test.test_compare_ragged_with_masks1_v2_function

  but it seems you are also testing some `v1_session` variant. Could you give me a hint how to run the Keras testing to also incude that variant so that I can experience the bug locally and fix the problem? Right now I am running
  ```
  bazel test --test_env=CUDA_VISIBLE_DEVICES=... --test_env=LD_LIBRARY_PATH=... --test_env=TF_FORCE_GPU_ALLOW_GROWTH=true keras/layers:recurrent_v2_test
  ```

Cheers & happy new Year!

**Comment by fchollet** [comment]
> but it seems you are also testing some v1_session variant. Could you give me a hint how to run the Keras testing to also incude that variant so that I can experience the bug locally and fix the problem?

You just need to add the following compilation flag to the bazel test command:

```
--define=tf_api_version=1
```

Please note that the issue only needs to be fixed if a previously-working V1 behavior is broken as a result of the PR. We are not adding new behaviors to V1, we just want to keep it working as it was. For new behaviors, it's fine not to test with V1 (and thus exclude the test).

**Comment by foxik** [comment]
Hi,

I tried 
- adding `--define=tf_api_version=1` to the above `bazel test` command, which unfortunately made no difference (still only v2_{eager,function} tests are run). 
- then I saw that `--config=v2` from `.bazelrc` from Keras root is also overriding it, so I also replaced lines
  ```
  build --config=v2
  test --config=v2
  ```
  by
  ```
  build --config=v1
  test --config=v1
  ```
  and run `bazel clean` followed by the above `bazel test` including the `--define=tf_api_version=1`; unfortunately, still only `v2_{eager,function}` tests.

Maybe I should be using a venv with TF 1 instead of the `tf-nightly` required by the `requirements.txt`? Or my Python 3.9 is too new? Or maybe the test should just be ignored somewhere in the battery of test that the TF1 configuration is running?

Sorry for not being much helpful...

**Comment by fchollet** [comment]
@qlzh727 Scott, do you know how contributors can run V1 tests?

I looked at the test logs and took a new look at the PR in this light. It seems that:

- The error occurs when running the test you added
- The error is caused by the weights of the recurrent layer (used in the test) not being initialized

I see the test was written with eager execution in mind (it expects that calling a layer on a constant returns a numpy-convertible tensor). So that would be the cause of the failure. As far as I can tell there is no existing (tested) behavior of V1 that is newly failing here, so you can simply exclude your new test from running in V1 and you'll be done. There are a few ways to do this, but the simplest is probably to add the decorator `@testing_utils.run_v2_only` to your test method.

Please try that and we'll run the test suite again.

**Comment by foxik** [comment]
Thanks a lot, I added the decorator :-)

**Review by fchollet — APPROVED** [review]
LGTM, thank you

**Comment by qlzh727** [comment]
> @qlzh727 Scott, do you know how contributors can run V1 tests?
> 
The tf_api_version flag will not work since the TF binary is imported from PIP and the flag is only consumed when we build TF. User might need to update the code with tf.compat.v1.disable_v2_behavior() to force the test run in v1.

> I looked at the test logs and took a new look at the PR in this light. It seems that:
> 
> * The error occurs when running the test you added
> * The error is caused by the weights of the recurrent layer (used in the test) not being initialized
> 
> I see the test was written with eager execution in mind (it expects that calling a layer on a constant returns a numpy-convertible tensor). So that would be the cause of the failure. As far as I can tell there is no existing (tested) behavior of V1 that is newly failing here, so you can simply exclude your new test from running in V1 and you'll be done. There are a few ways to do this, but the simplest is probably to add the decorator `@testing_utils.run_v2_only` to your test method.
> 
> Please try that and we'll run the test suite again.

**Comment by fchollet** [comment]
PR merged. Thank you for the contribution! 👍

**Comment by foxik** [comment]
Thanks for the guidance :-)

I took the liberty in creating a cherry-pick for r2.8 in #15862. I understand it might be too late, but I believe the benefit (large speedup on GPU) is considerable, given that we just reuse an existing cuDNN implementation, which is used for masked tensors.
